### DISCUSSION TITLE
How to configure ant design v5 for micro frontend project?

### DISCUSSION THREAD

**Discussion by pooryak** [ant-design/ant-design#40409] [Q&A] [answered]
I have design system packages and I want to build them with ant design v5. I wonder that Should I put a theme wrapper on the whole project and have a token provider for each piece of my design system?

**Comment by arvinxx** [comment, marked as answer]
You can use v5 without any wrapper in micro frontend if you don't override style. If you overide, maybe it's better to change prefix

**Reply by pooryak** [reply]
And if I want use antd v5 with tailwind I should export theme from antd and give it to tailwind?

**Reply by arvinxx** [reply]
maybe you can use 
```ts
import { theme } from 'antd/lib';

const tokens = theme.darkAlgorithm({ ... your seed token });

module.exports = {
  theme: {
    screens: {
      sm:  token.screenXSMin,
    },
    colors: {
      'blue': token.blue,
      'purple': token.purple,
      ...
    },
}

```

seeds params: https://ant.design/docs/react/customize-theme-cn#seedtoken

**Comment by madiyetov** [comment]
Hey @arvinxx, we use `<ConfigProvider prefixCls={val} iconPrefixCls={val} />` in combination with `<StyleProvider container={container}>` in all remote microfrontends, but some of the ant's styles aren't loading when asyncronously loading another microfrontend. So we've decided to make `@ant-design/cssinjs` a singleton, but it generated another error:
```
React Router caught the following error during render TypeError: Cannot read properties of undefined (reading '_tokenKey')
    at useStyleRegister (index.js:269:24)
    at useResetIconStyle (useResetIconStyle.js:28:79)
    at ProviderChildren (index.js:223:53)
    at renderWithHooks (react-dom.development.js:2719:157)
    at mountIndeterminateComponent (react-dom.development.js:3293:1445)
    at beginWork (react-dom.development.js:3632:93)
    at beginWork$1 (react-dom.development.js:4688:93)
    at performUnitOfWork (react-dom.development.js:4514:150)
    at workLoopSync (react-dom.development.js:4500:30)
    at renderRootSync
```

Would be great to have a documentation about using antd in microfrontend architecture.

**Reply by arvinxx** [reply]
@MadCcc

**Reply by madiyetov** [reply]
Hey @MadCcc, let me explain the current situation. We have basic host-remote microfrontend architecture using webpack's `ModuleFederationPlugin`. Where the host doesn't know anything about `antd`. So the simple architecture looks something like this:

<img width="1231" alt="image" src="https://github.com/ant-design/ant-design/assets/17025375/d07e42a0-e2b2-4e22-9e5b-19fc3999d4af">

(Here, `StyleProvider` is used to pass `shadow dom` element to it, where the `antd` styles will be injected.)
In this case, we get issues like erasing `style` tags for components that are used in both projects. And also the error that was mentioned above.

So I've done a little research and found that `StyleProvider` uses `StyleContext` which creates `cache` and `counter`s for tokens. And it looked a lot complicated. So I've decided to move `StyleProvider` and `ConfigProvider` to the host project and put all remotes under their context. Also make `@ant-design/cssinjs` a singleton. Like this: 

<img width="1231" alt="image" src="https://github.com/ant-design/ant-design/assets/17025375/fe71b996-49f2-43e4-bd18-53e4394d7f3a">


After these changes, everything is working fine. My first guess is that `antd` is designed for standalone applications and doesn't take into account cases like microfrontend. Or not? Could you please explain what can be the problem in this solution and what would be the best solution with this architecture?

Thanks!
